pkg/codegen/go/utilities.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gen
+
+import (
+	"strings"
+	"unicode"
+)
+
+// isReservedWord returns true if s is a Go reserved word as per
+// https://golang.org/ref/spec#Keywords
+func isReservedWord(s string) bool {
+	switch s {
+	case "break", "default", "func", " interface", "select",
+		"case", "defer", "go", "map", "struct",
+		"chan", "else", "goto", "package", "switch",
+		"const", "fallthrough", "if", "range", "type",
+		"continue", "for", "import", "return", "var":
+		return true
+
+	default:
+		return false
+	}
+}
+
+// isLegalIdentifierStart returns true if it is legal for c to be the first character of a Go identifier as per
+// https://golang.org/ref/spec#Identifiers
+func isLegalIdentifierStart(c rune) bool {
+	return c == '_' || unicode.In(c, unicode.Letter)
+}
+
+// isLegalIdentifierPart returns true if it is legal for c to be part of a Go identifier (besides the first character)
+// https://golang.org/ref/spec#Identifiers
+func isLegalIdentifierPart(c rune) bool {
+	return c == '_' ||
+		unicode.In(c, unicode.Letter, unicode.Digit)
+}
+
+// makeValidIdentifier replaces characters that are not allowed in Go identifiers with underscores. A reserved word is
+// prefixed with _. No attempt is made to ensure that the result is unique.
+func makeValidIdentifier(name string) string {
+	var builder strings.Builder
+	firstChar := 0
+	for i, c := range name {
+		// ptr dereference
+		if i == 0 && c == '&' {
+			firstChar++
+		}
+		if i == firstChar && !isLegalIdentifierStart(c) || i > 0 && !isLegalIdentifierPart(c) {
+			builder.WriteRune('_')
+		} else {
+			builder.WriteRune(c)
+		}
+	}
+	name = builder.String()
+	if isReservedWord(name) {
+		return "_" + name
+	}
```

Comment:
Should move this logic to the beginning of the function and conditionally add the underscore to the builder.

Reply:
By definition characters in the reserved keyword list for Go must be legal for use in the language. Ultimately we'd end up with the same code:
```go
if isReservedWord(name) {
		return "_" + name
	}
```

just moved to the front of the function. What's the benefit? Short-circuiting the rest of the logic I suppose?

Reply:
I think it's more efficient to do as part of the builder than using the string concatenation. Not a huge deal.